plugin/storage/es/spanstore/reader.go

```diff
@@ -205,39 +205,80 @@ func (s *SpanReader) FindTraces(traceQuery *spanstore.TraceQueryParameters) ([]*
 }
 
 func (s *SpanReader) multiRead(traceIDs []string, startTime, endTime time.Time) ([]*model.Trace, error) {
+
 	if len(traceIDs) == 0 {
 		return []*model.Trace{}, nil
 	}
 	searchRequests := make([]*elastic.SearchRequest, len(traceIDs))
 
-	for i, traceID := range traceIDs {
-		query := elastic.NewTermQuery("traceID", traceID)
-		searchRequests[i] = elastic.NewSearchRequest().IgnoreUnavailable(true).Type("span").
-			Source(elastic.NewSearchSource().Query(query).Size(defaultDocCount))
-	}
-
+	var traces []*model.Trace
 	// Add an hour in both directions so that traces that straddle two indexes are retrieved.
-	// ie starts in one and ends in another.
+	// i.e starts in one and ends in another.
 	indices := findIndices(spanIndexPrefix, startTime.Add(-time.Hour), endTime.Add(time.Hour))
 
-	results, err := s.client.MultiSearch().
-		Add(searchRequests...).
-		Index(indices...).
-		Do(s.ctx)
-	if err != nil {
-		return nil, err
-	}
+	var nextTime uint64
 
-	var traces []*model.Trace
-	for _, result := range results.Responses {
-		if result.Hits == nil || len(result.Hits.Hits) == 0 {
-			continue
+	searchAfterTime := make(map[string]uint64)
+	totalDocumentsFetched := make(map[string]int)
+	tracesMap := make(map[string]*model.Trace)
+	for {
+		if traceIDs == nil || len(traceIDs) == 0 {
+			break
 		}
-		spans, err := s.collectSpans(result.Hits.Hits)
+
+		for i, traceID := range traceIDs {
+			query := elastic.NewTermQuery("traceID", traceID)
+			if val, ok := searchAfterTime[traceID]; ok {
+				nextTime = val
+			}
+			if nextTime == 0 {
+				searchRequests[i] = elastic.NewSearchRequest().IgnoreUnavailable(true).Type("span").Source(elastic.NewSearchSource().Query(query).Size(defaultDocCount).Sort("startTime", true))
+			} else {
+				searchRequests[i] = elastic.NewSearchRequest().IgnoreUnavailable(true).Type("span").Source(elastic.NewSearchSource().Query(query).Size(defaultDocCount).Sort("startTime", true).SearchAfter(nextTime))
+			}
+			}
+		// set traceIDs to empty
+		traceIDs = nil
+		results, err := s.client.MultiSearch().Add(searchRequests...).Index(indices...).Do(s.ctx)
+
 		if err != nil {
 			return nil, err
 		}
-		traces = append(traces, &model.Trace{Spans: spans})
+
+		if results.Responses == nil || len(results.Responses) == 0 {
+			break
+		}
+
+		for _, result := range results.Responses {
+			if result.Hits == nil || len(result.Hits.Hits) == 0 {
+				continue
+			}
+			spans, err := s.collectSpans(result.Hits.Hits)
+			if err != nil {
+				return nil, err
+			}
+			lastSpan := spans[len(spans)-1]
+			lastSpanTraceID := lastSpan.TraceID.String()
+
+			if traceSpan, ok := tracesMap[lastSpanTraceID]; ok {
+				for _, span := range spans {
+					traceSpan.Spans = append(traceSpan.Spans, span)
```

Comment:
Is the order of this map important? If not, we might want to consider doing some parallel processing here, no?

Reply:
@jpkrohling l don't think so. I was thinking about concurrent processing ,like run the elastic-search query separate and send the results in a channel to construct the map ?

Reply:
That's what I had in mind, yes.

Reply:
@jpkrohling did more thinking on this. elastic search query is dependent on the results of the query to get the start time that'll be used for next search after query. I don't think concurrency might work here. That being said, but, I will give more thoughts on this weekend.